repentogon/resources/scripts/main_ex.lua

```diff
@@ -1407,6 +1412,109 @@ function _RunPostPickupSelection(callbackID, param, pickup, variant, subType, ..
 	return recentRet
 end
 
+-- Custom handling for MC_PRE_STATUS_EFFECT_APPLY
+-- Handle type checking here since the callback is called for several status effects with different types,
+-- terminate early if false is returned,
+-- and use additive callback logic instead of the default one.
+local function RunPreStatusEffectApplyCallback(callbackID, param, status, entity, entityRef, duration, extraParam1, extraParam2)
+	local recentRet = nil
+
+	for callback in GetCallbackIterator(callbackID, param) do
+		local ret = RunCallbackInternal(callbackID, callback, status, entity, entityRef, duration, extraParam1, extraParam2)
+
+		if type(ret) == "boolean" then
+			if ret == false then return false end
+			recentRet = ret
+		elseif type(ret) == "table" then
+			if status == StatusEffect.CONFUSION then
```

Comment:
This section can probably be refactored a bit. I believe that you can generalize the logic with a table to get the return types, and also manually create & call a typecheck function here, like:
```lua
local PreStatusApplyTypes = {
  [StatusEffect.KNOCKBACK] = {"integer", "Vector", "boolean"}
  ...
}

...

local err = checkTableTypeFunction(PreStatusApplyTypes[status])
if err then
  logError(callbackID, callback.Mod.Name, err)
end
```

Reply:
On second thought, you could just put the functions in a table, haha
```lua
local PreStatusApplyTypeCheckFunctions = {
  [StatusEffect.KNOCKBACK] = checkTableTypeFunction({"integer", "Vector", "boolean"}),
  ...
}
```